### PR TITLE
feat: Allow numeric generics to non inlined ACIR functions

### PR DESCRIPTION
**Pull request by vezenovm**
# Description

## Problem\*

Resolves https://github.com/noir-lang/noir/issues/4826. 

## Summary\*

When running `nargo info` with the following code:
```rust
use dep::std::hash::{pedersen_hash_with_separator, poseidon2::Poseidon2};

global NUM_HASHES = 2;
global HASH_LENGTH = 10;

#[fold]
pub fn poseidon_hash<N>(inputs: [Field; N]) -> Field {
    Poseidon2::hash(inputs, inputs.len())
}
fn main(
    to_hash: [[Field; HASH_LENGTH]; NUM_HASHES],
    enable: [bool; NUM_HASHES]
) -> pub [Field; NUM_HASHES] {
    let mut result = [0; NUM_HASHES];
    for i in 0..NUM_HASHES {
        let enable = enable[i];
        let to_hash = to_hash[i];
        if enable {
            result[i] = poseidon_hash(to_hash);
        }
    }
    result
}
```
Screenshot 2024-04-17 at 3 46 54 PM

When using `pedersen_hash` we also get 10 opcodes for `main`. 

Once we get more granularity in our inline types (fold vs. inline for proving) we can most likely fully resolve this issue (https://github.com/noir-lang/noir/issues/4688).

## Documentation\*

Check one:
- [ ] No documentation needed.
- [ ] Documentation included in this PR.
- [ ] **[For Experimental Features]** Documentation to be submitted in a separate PR.

# PR Checklist\*

- [ ] I have tested the changes locally.
- [ ] I have formatted the changes with [Prettier](https://prettier.io/) and/or `cargo fmt` on default settings.

### CODE REVIEW

File: compiler/noirc_frontend/src/hir/resolution/resolver.rs [hunk 1]
```diff
@@ -929,13 +929,14 @@ impl<'a> Resolver<'a> {
...
         let attributes = func.attributes().clone();
+        let should_fold = attributes.is_foldable();
 
         let mut generics = vecmap(&self.generics, |(_, typevar, _)| typevar.clone());
         let mut parameters = vec![];
         let mut parameter_types = vec![];
 
         for Param { visibility, pattern, typ, span: _ } in func.parameters().iter().cloned() {
-            if visibility == Visibility::Public && !self.pub_allowed(func) {
+            if visibility == Visibility::Public && !self.pub_allowed(func) && !should_fold {
```
Comment (vezenovm) [inline compiler/noirc_frontend/src/hir/resolution/resolver.rs:939]: We do not want this warning as we will want to be able to specify public inputs on foldable functions

File: compiler/noirc_frontend/src/hir/type_check/mod.rs [hunk 2]
```diff
@@ -172,7 +172,9 @@ fn check_if_type_is_valid_for_program_input(
     errors: &mut Vec<TypeCheckError>,
 ) {
     let meta = type_checker.interner.function_meta(&func_id);
-    if (meta.is_entry_point || meta.should_fold) && !param.1.is_valid_for_program_input() {
+    if (meta.is_entry_point && !param.1.is_valid_for_program_input())
+        || (meta.should_fold && !param.1.is_valid_entry_point_input())
```
Comment (vezenovm) [inline compiler/noirc_frontend/src/hir/type_check/mod.rs:176]: I was unsure if I should trigger a different error here as it is technically still true that we cannot accept anything but sized types to the entry point.

Comment (jfecher) [reply compiler/noirc_frontend/src/hir/type_check/mod.rs:176]: How does this work for generic code? Where are the numeric generics instantiated with the desired sizes if they're in an entry point?

File: compiler/noirc_frontend/src/monomorphization/mod.rs [hunk 3]
```diff
@@ -283,7 +283,13 @@ impl<'interner> Monomorphizer<'interner> {
...
 
         let meta = self.interner.function_meta(&f).clone();
-        let func_sig = meta.function_signature();
+        let mut func_sig = meta.function_signature();
+        // Follow the bindings of the function signature for entry points
+        // which are not `main` such as foldable functions.
+        for param in func_sig.0.iter_mut() {
+            param.1 = param.1.follow_bindings();
+        }
+        func_sig.1 = func_sig.1.map(|return_type| return_type.follow_bindings());
```
Comment (jfecher) [inline compiler/noirc_frontend/src/monomorphization/mod.rs:292]: Were you getting errors without this?

File: compiler/noirc_frontend/src/hir/resolution/resolver.rs (hunk 1 above)
Comment (jfecher) [reply compiler/noirc_frontend/src/hir/resolution/resolver.rs:939]: `should_fold` should just be included as part of `pub_allowed` it seems. It looks like each place pub_allowed is used so is folding.

**Comment by vezenovm** [comment]
> How does this work for generic code? Where are the numeric generics instantiated with the desired sizes if they're in an entry point?

They are instantiated during monomorphization. I wasn't sure the best name for the `is_valid_entry_point_input` function, as they are an "ACIR function" which acts similar to an entry point, but they are not an "ACIR program entry point" like `main`. I was just trying to distinguish with `is_valid_for_program_input`. Maybe it should just be `is_valid_non_inlined_function_input`?

**Comment by vezenovm** [comment]
> should_fold should just be included as part of pub_allowed it seems. It looks like each place pub_allowed is used so is folding.

Yup you are correct I moved that check into `pub_allowed`

File: compiler/noirc_frontend/src/hir/type_check/mod.rs (hunk 2 above)
Comment (vezenovm) [reply compiler/noirc_frontend/src/hir/type_check/mod.rs:176]: They are instantiated during monomorphization. I wasn't sure the best name for the is_valid_entry_point_input function, as they are an "ACIR function" which acts similar to an entry point, but they are not an "ACIR program entry point" like main. I was just trying to distinguish with is_valid_for_program_input. Maybe it should just be is_valid_non_inlined_function_input?

Comment (jfecher) [reply compiler/noirc_frontend/src/hir/type_check/mod.rs:176]: Yeah, I think it'd benefit from a different name. Maybe we should just call out that it is for folding specifically.

Comment (vezenovm) [reply compiler/noirc_frontend/src/hir/type_check/mod.rs:176]: I have the make the issue still but I am planning to add an `inline(never)` flag which will also be used here

Comment (vezenovm) [reply compiler/noirc_frontend/src/hir/type_check/mod.rs:176]: `is_valid_non_inlined_function_input` is a bit of a mouthful but it is the most accurate

**Review by jfecher — APPROVED** [review]

**Pull request merged**